asio/include/xtt/asio/context.inl

```diff
@@ -0,0 +1,340 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ * 
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+namespace xtt {
+namespace asio {
+
+    template <typename GPKLookupCallback,
+              typename AssignIdCallback,
+              typename SuccessCallback,
+              typename ErrorCallback>
+    void
+    context::run_state_machine(return_code current_rc,
+                               uint16_t bytes_requested,
+                               unsigned char *io_ptr,
+                               GPKLookupCallback gpk_lookup_callback,
+                               AssignIdCallback assign_id_callback,
+                               SuccessCallback success_callback,
+                               ErrorCallback error_callback)
+    {
+        switch (current_rc) {
+            case return_code::WANT_WRITE:
+                {
+                    boost::asio::async_write(socket_,
+                                             boost::asio::buffer(io_ptr,
+                                                                 bytes_requested),
+                                             boost::asio::bind_executor(strand_,
+                                                                        [this, gpk_lookup_callback, assign_id_callback, success_callback, error_callback]
+                                                                        (auto&& ec, auto&& bytes_transferred)
+                                                                        {
+                                                                            if (ec) {
+                                                                                error_callback(return_code::BAD_IO);
```

Comment:
Shouldn't just pass the actual `ec` from the `async_write` call to the completion handler?  Why hide the actual error with the less specific and probably less useful `BAD_IO`?